Pagination implemented in erlcloud_ec2

SergeyOvchinnik opened the pull request:
Problem: 
7 erlcloud_ec2 describe functions with pagination implemented on API level did not support pagination feature on library level

Solution:
Implemented pagination feature for all 7 functions
Implemented unit tests for them as well

Reviewers:
@ddosia @motobob

SergeyOvchinnik committed:
Implemented pagination support in erlcloud_ec2

Implemented pagination in the following functions:
- describe_instances
- describe_instance_status
- describe_reserved_instances_offerings
- describe_snapshots
- describe_spot_price_history
- describe_spot_fleet_instances
- describe_tags

Added pagination unit tests to erlcloud_ec2_tests

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -162,15 +167,18 @@
+         %% Customer gateways
+         describe_customer_gateways/0, describe_customer_gateways/1, describe_customer_gateways/2, describe_customer_gateways/3,
+
+         describe_tags_tester/0
```
should not be here

SergeyOvchinnik replied:
Removed

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,36 +1355,138 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+
+-define(INSTANCES_MR_MIN, 5).
+-define(INSTANCES_MR_MAX, 1000).
+-define(INSTANCES_MR_DEFAULT, 1000).
```
why do we need a second define for the same value?

SergeyOvchinnik replied:
Removed

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1941,46 +2051,187 @@ describe_snapshot_attribute(SnapshotID, create_volume_permission, Config)
-    describe_snapshots(SnapshotIDs, none, none, default_config()).
+-define(SNAPSHOTS_MR_MIN, 5).
+-define(SNAPSHOTS_MR_MAX, 1000).
+-define(SNAPSHOTS_MR_DEFAULT, 1000).
```
same. why do we need to default it several times? 
if all EC2 pagination uses same number please use the same define

SergeyOvchinnik replied:
Some functions have different max and min values, I'll see if I can optimise the defines once I've implemented all 7 functions

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1941,46 +2051,187 @@ describe_snapshot_attribute(SnapshotID, create_volume_permission, Config)
+    when is_list(Owner), is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX, is_list(NextToken) ->
+    describe_snapshots(Owner, none, MaxResults, NextToken, default_config());
+describe_snapshots(MaxResults, NextToken, Config)
+    when is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX, is_list(NextToken), is_record(Config, aws_config) ->
```
how about having a guard for `MaxResults` once in wides fun and not in every fun?

SergeyOvchinnik replied:
Agree, removed

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1941,46 +2051,187 @@ describe_snapshot_attribute(SnapshotID, create_volume_permission, Config)
+         is_integer(MaxResults), MaxResults >= ?SNAPSHOTS_MR_MIN, MaxResults =< ?SNAPSHOTS_MR_MAX,
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    Params = [{"Owner", Owner}, {"RestorableBy", RestorableBy}, {"MaxResults", MaxResults}, {"NextToken", NextToken}],
```
please generalise pagination work with `NextToken`/`MaxResult`

SergeyOvchinnik replied:
I don't really understand what you mean by that. Do you want me to create a separate function which would make paginated calls for any specified action? And make all paginated interfaces call it instead of doing the pagination logic themselves?

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1941,46 +2051,187 @@ describe_snapshot_attribute(SnapshotID, create_volume_permission, Config)
+                    [] ->
+                        undefined;
+                    NTxml ->
+                        element(5,lists:nth(1,element(9,lists:nth(1,NTxml))))
```
? what does it do?

SergeyOvchinnik replied:
It extracts the Next Token from the response. I've moved it to a separate function

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -2938,46 +3189,38 @@ delete_tags(ResourceIds, TagsList, Config) when is_list(ResourceIds)->
+
+-define(TAGS_MR_MIN, 5).
+-define(TAGS_MR_MAX, 1000).
+-define(TAGS_MR_DEFAULT, 1000).
```
same comment about max

SergeyOvchinnik replied:
Also removed

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -2987,15 +3230,113 @@ describe_tags(Filters, Config) ->
     end.
 
--spec filter_name(filter_name()) -> string().
+lists_are_the_same(List1, List2) ->
```
belongs to tests

SergeyOvchinnik replied:
My mistake, forgot to remove

motobob commented on test/erlcloud_ec2_tests.erl:
```diff
@@ -720,3 +720,284 @@ describe_images_tests(_) ->
+.
+
+describe_instances_pagination_tester(MaxInstances, PageSize) ->
+    meck:new(erlcloud_aws, [passthrough]),
```
as discussed before please meck it once the suite and 10 times per each test

motobob commented on test/erlcloud_ec2_tests.erl:
```diff
@@ -720,3 +720,284 @@ describe_images_tests(_) ->
+    meck:unload(erlcloud_aws),
+    ?assert(lists_are_the_same(AllInstances, PagedInstances)).
+
+describe_snapshots_pagination_tester(MaxSnapshots, PageSize) ->
```
copy-pasta detected. please remove&optimise

SergeyOvchinnik replied:
Optimised slightly

motobob reviewed (changes requested):
please address the comments

SergeyOvchinnik commented:
@motobob I've made the changes we discussed. I have not implemented the generic describe_all function yet due to its complexity, I suggest we add that after pagination is implemented in all 7 paginatable functions

motobob commented:
if you feel it will make sense to do 7 copy/pastes and then optimise it into one fun I'm fine but the PR will not get merged until its generalised.

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -3285,3 +3597,12 @@ extract_cgw(Node) ->
  ].
+
+-spec extract_next_token(list()) -> string() | undefined.
+extract_next_token(NTxml) ->
```
single case fun should be claused.

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -3285,3 +3597,12 @@ extract_cgw(Node) ->
+        [] ->
+            undefined;
+        NTxml ->
+            element(5,lists:nth(1,element(9,lists:nth(1,NTxml))))
```
can this fail with bararg?

SergeyOvchinnik commented:
@motobob I have generalised implementation of pagination with describe_all function as you suggested and made some other small changes

zdaniel86 commented:
Guys! http://erlang.org/doc/man/httpc.html supports streaming of HTTP body with flow control via `{stream, {self, once}}` and `http:stream_next/1`. So we can stream even not paginated AWS API. The only thing should be implemented is JSON parser of top level elements in the list. Once we fetch the binary of single JSON element from the stream it can be passed to usual JSON parser. It seems even simpler and unified than paginated API. What do you think?

nalundgaard commented:
erlcloud doesn't use httpc.

motobob commented:
@zdaniel86 how is streaming of client which is not used in the library related to problem of a missing API? 
This PR addresses the API portion.

zdaniel86 commented:
I meant that it is general approach which can be used regardless of pagination support on server side. `lhttpc` has similar functionality.

SergeyOvchinnik commented:
@motobob @ddosia Alright, so I addressed all comments and implemented pagination in 6 functions. I can't really implement pagination in `describe_spot_fleet_instances` the same way without ruining backward compatibility since pagination is already implemented there in a slightly different way. What should I do about it?

SergeyOvchinnik commented:
Nevermind, implementing it now with full backward compatibility

ddosia commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1365,117 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+% Function for describing all instances page-by-page
+% Default page size is 1000
+-spec describe_instances_all() -> ec2_normal_describe_return().
+describe_instances_all() ->
```
I still see `describe_instances_all()` as confusing, since you can get all instances through `describe_instances()`, isn't it? cc @motobob

motobob replied:
the point and the difference of *_all() is to provide a wrapper around paginated API to still get all the things but in chunks. 
this exist in other modules and added for consistency only. 
the might be unnecessary complication but well let's discuss.

ddosia commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,27 +1538,131 @@ extract_block_device_mapping_status(Node) ->
+describe_instance_status(MaxResults)
+    when is_integer(MaxResults) ->
+    describe_instance_status([], [], MaxResults, undefined, default_config()).
+
```
What I would do is to introduce a type:

``` erlang
-type ok_error(Ok) :: {ok, Ok} | {error, term()}.
```

And then get rid of `ec2_normal_describe_return()` and `ec2_paged_describe_return()` in favour of more clear:

``` erlang
-describe_instance_status( ec2_param_list(), filter_list()
                                           | aws_config()
                                           | ec2_max_result() ) -> ok_error(proplist());
                         ( ec2_max_result(), aws_config()
                                           | ec2_token() ) -> ok_error({proplist(), ec2_token()}).
```

and I actually insist that `{ok, {proplist(), ec2_token()}}` would be more convenient than `{ok, proplist(), ec2_token()}`.

motobob replied:
agree to the first part and disagree to the last part with more tuple wrapping. what for? and again rest of the modules have it with tuple of three elements so let's not produce more snowflakes.

SergeyOvchinnik replied:
Ok, I'll implement a type for errors then and use it everywhere

ddosia commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,27 +1538,131 @@ extract_block_device_mapping_status(Node) ->
         {ok, Doc} ->
             Path = "/DescribeInstanceStatusResponse/instanceStatusSet/item",
             {ok, [ extract_instance_status(Item) || Item <- xmerl_xpath:string(Path, Doc) ]};
-        {error, _} = Error ->
```
Why did you remove this line? Generally speaking it is more efficient. And readable (imo).

SergeyOvchinnik replied:
Agree, I'll replace all error cases with that

ddosia commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,27 +1538,131 @@ extract_block_device_mapping_status(Node) ->
--spec(describe_instance_status/3 :: (proplist(), filter_list(), aws_config()) -> proplist()).
-describe_instance_status(Params, Filter, Config) ->
+-spec describe_instance_status() -> ec2_normal_describe_return().
+describe_instance_status() ->
```
There was no `describe_instance_status/0` before? What does it do? My guess it will describe statuses of all instances? So what `describe_instance_status_all/...` does?

ddosia commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,27 +1538,131 @@ extract_block_device_mapping_status(Node) ->
+    when is_integer(MaxResults), is_list(NextToken) ->
+    describe_instance_status([], [], MaxResults, NextToken, default_config()).
+
+-spec describe_instance_status(ec2_param_list(), filter_list(), aws_config()) -> ec2_normal_describe_return();
```
I am sorry, I don't believe that it the best way to design api. There was clear `describe_instance_status(proplist(), filter_list(), aws_config())` where you always had config as a 3rd argument. But after these changes we basically can pass `ec2_max_result() | ec2_token()` instead (which implies default config?).

To me it looks like a mess, to be honest. Can we leave it as is and add 4th arg for tokens? Or if some of the args are mutually exclusive reflect it.

SergeyOvchinnik replied:
Ok, lets make it clear. If one wants to specify `Params`, `Filter`, and page size `MaxResults` and use default config then what call would they need to make instead of `describe_instance_status(Params, Filter, MaxResults)` ?

ddosia reviewed (changes requested):
As a summary I would highlight 3 points which have room for improvement:
- data types could more clear;
- `describe_..._all` is unnecessary in my opinion;
- supporting all possible combination of arguments passage does not improve readability and is error prone.

SergeyOvchinnik commented:
I removed all unnecessary clauses in describe functions, all pagination functionality is now implemented in couple clauses in the end of every function and can be accessed by specifying some default values in the call. I also decided to restrict user to always having to specify both MaxResults and NextToken together and never separately in normal calls. All of this also made the functions a lot shorter. I decided not to touch describe_spot_fleet_instances since pagination is already implemented there and implementing an alternative while maintaining backward compatibility will just create more mess and no extra functionality.

SergeyOvchinnik commented:
I can confirm that in DescribeInstances calls NextToken only stores the index of the element page starts with and does not store reference to the original call or any of its parameters. Thanks a lot AWS. To be fair, it kind of makes sense to me why AWS does it that way but I don't understand why they don't just use an integer instead of a gazillion symbol string if that's the case.

SergeyOvchinnik commented:
As for loss of some results between paginated calls: I don't think it's a real issue. True, if something changes while the _all function is making its calls, some of the results it already retrieved may become outdated. And if we test it the same way I used to test it manually: make normal call, then make paginated calls and compare results, then surely there is a chance something might change between those calls and results will not turn out to be the same. But I can reproduce the same problem by just making 2 normal calls and not using pagination at all.

SergeyOvchinnik commented:
@ddosia 
I have generalised big parts of all describe functions : making call to AWS, extracting results from Doc, and extracting NextToken from Doc. Now there is only one ec2_query call written for all of those 7 functions, only one list comprehension extracting all results and only one way to extract next token. 

The next step could be to use the new ec2_describe_query to write a very short implementation of describe_all and get rid of that 80 line mess. Do you agree with such approach?

SergeyOvchinnik commented:
I've appended this pr with last changes, code ready to merge

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
 attribute_xpath(_, AttributeName) ->
     "/DescribeInstanceAttributeResponse/" ++ AttributeName ++ "/value".
 
+-type ec2_instances_ids() :: [string()].
```
just a note that `erlcloud_aws:param_list()` work with both strings and binaries but that;s is fine to set it this way

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -337,7 +354,8 @@ extract_volume_status(Node) ->
     authorize_security_group_ingress(GroupName, IngressSpec, default_config()).
 
--spec(authorize_security_group_ingress/3 :: (string(), ec2_ingress_spec() | [ vpc_ingress_spec() ], aws_config()) -> ok | {error, any()}).
+-spec(authorize_security_group_ingress/3 :: (string(), ec2_ingress_spec() | [ vpc_ingress_spec() ], aws_config())
```
need to rebase over upstream to pick up R19 specs format

motobob replied:
this still has ::

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1039,7 +1068,10 @@ describe_addresses(PublicIPs, Config)
             Items = xmerl_xpath:string("/DescribeAddressesResponse/addressesSet/item", Doc),
-            {ok, [[{public_ip, get_text("publicIp", Item)}, {instance_id, get_text("instanceId", Item, none)}] || Item <- Items]};
+            {ok, [[
+                    {public_ip, get_text("publicIp", Item)},
```
so much unnecessary changes. for whitespaces

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
--spec(describe_instances/2 :: ([string()], filter_list() | aws_config()) -> proplist()).
+-spec describe_instances(ec2_instances_ids(), filter_list()) -> ok_error(proplist());
+    (ec2_instances_ids(), aws_config()) -> ok_error(proplist());
+    (ec2_max_result(), ec2_token()) -> ok_error(proplist(), ec2_token()).
```
wrong spec?

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+         is_integer(MaxResults) andalso MaxResults >= ?INSTANCES_MR_MIN andalso MaxResults =< ?INSTANCES_MR_MAX,
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    Params = [{"MaxResults", MaxResults}, {"NextToken", NextToken}] ++ list_to_ec2_filter(Filter),
```
what is the point of passing NExtToken when that is `undefined` ?

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+    Params = [{"MaxResults", MaxResults}, {"NextToken", NextToken}] ++ list_to_ec2_filter(Filter),
+    case ec2_query(Config, "DescribeInstances", Params, ?NEW_API_VERSION) of
+        {ok, Doc} ->
+            Reservations = extract_results("DescribeInstances", "reservationSet", fun extract_reservation/1, Doc),
```
I'm not a fan of idea inside the function regarding `++"Response/"`
please update it all to only append `"/"`

```
extract_results("DescribeInstancesReponse", "reservationSet", fun extract_reservation/1, Doc),
```

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+    case ec2_query(Config, "DescribeInstances", Params, ?NEW_API_VERSION) of
+        {ok, Doc} ->
+            Reservations = extract_results("DescribeInstances", "reservationSet", fun extract_reservation/1, Doc),
+            NewNextToken = extract_next_token("DescribeInstances", Doc),
```
same here, please make it so that
s/DescribeInstances/DescribeInstancesReponse

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1347,38 +1384,93 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+                            (ec2_max_result(), aws_config()) -> ok_error(proplist());
+                            (filter_list(), aws_config()) -> ok_error(proplist()).
+describe_instances_all(Param1, Param2) ->
+    describe_all(describe_instances, Param1, Param2).
```
i'm not fan of passing atoms and then using apply. 
why not pass `fun describe_intances/2`.
this way it's clear which arity of the fun is expected to be run

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,27 +1535,72 @@ extract_block_device_mapping_status(Node) ->
+
+-spec describe_instance_status_all(ec2_param_list(), filter_list()) -> ok_error(proplist()).
+describe_instance_status_all(Param1, Param2) ->
+    describe_all(describe_instance_status, Param1, Param2).
```
well...let's remove this boiler plate and leave only 
describe_all/2 describe_all/3 exported so that library consumers can use it like

```
    describe_all(fun erlcloud:describe_instances/2, Params, Filter),
   describe_all(fun erlcloud:describe_instances/3, Params, Filter, AWSConfig),
```

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1736,45 +1885,75 @@ describe_reserved_instances_offerings(Selector, Config)
+            Offerings = extract_results("DescribeReservedInstancesOfferings", "reservedInstancesOfferingsSet",
+                                        fun extract_reserved_instances_offering/1, Doc),
+            NewNextToken = extract_next_token("DescribeReservedInstancesOfferings", Doc),
+            case NewNextToken of
```
why rest of the fun do not expose this case behaviour? let's be uniform across the whole module

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1934,55 +2113,114 @@ describe_snapshot_attribute(SnapshotID, Attribute) ->
-        {error, _} = Error ->
-            Error
+    when is_list(SnapshotIDs),
+    is_list(Owner) orelse Owner =:= none,
```
spaces alignments please or do not change it

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -3281,7 +3634,101 @@ extract_cgw(Node) ->
+describe_all(DescribeFunction, Config)
+    when is_record(Config, aws_config) ->
+    describe_all(DescribeFunction, ?DEFAULT_MR, Config);
+describe_all(DescribeFunction, Param1) ->
```
can a Param1 be an integer? if so it clashes

motobob reviewed (changes requested):
need next round

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -3281,7 +3634,101 @@ extract_cgw(Node) ->
+    describe_all(DescribeFunction, Param1, ?DEFAULT_MR, default_config()).
+
+-spec describe_all(atom(), any(), any()) -> ok_error(proplist()).
+describe_all(DescribeFunction, MaxResults, Config)
```
well...after reading all the clauses of this supper generic function i would agree with @ddosia that is does not belong to the library.
Therefore, sorry for waste of your time but let's remove it from the lib and move it you your application level, obviously nuking rest of `_all`.

motobob reviewed (changes requested):
no more _all.

motobob commented on src/erlcloud_aws.erl:
```diff
@@ -236,7 +237,15 @@ aws_request_form_raw(Method, Scheme, Host, Port, Path, Form, Headers, Config) ->
     end.
 
+-spec single_param(term(), string()) -> proplist().
+single_param(undefined, _ParamName) ->
```
how is this different from `param_list(undefined,_Key)`?
please remove

SergeyOvchinnik replied:
Removing single_param completely

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1441,25 +1507,59 @@ extract_block_device_mapping_status(Node) ->
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    AllParams = Params ++
+                erlcloud_aws:single_param(MaxResults, "MaxResults") ++
```
param_list is fine here.

SergeyOvchinnik replied:
Param_list seems to be fine with both MaxResults and NextToken everywhere, seems like API deals with MaxResult and MaxResult.N formats equaly well. But some other params fail one of the tests in erlcloud_ec2_tests if they are used in that format. I might just remove single_param function and turn all single params back into [{"MaxResult", 5}] format for now

motobob commented on src/erlcloud_ec2.erl:
```diff
@@ -1765,13 +1907,13 @@ describe_reserved_instances_offerings_all(Selector, Config) ->
-        {ok, Res, NextToken} ->
-            NewSelector = [{token, NextToken} | proplists:delete(token, Selector)],
-            describe_reserved_instances_offerings_all(NewSelector, Config, lists:append(Acc, Res));
+            {ok, Acc ++ Res};
```
what is the purpose of this change?

SergeyOvchinnik replied:
Consistency

motobob reviewed (changes requested):
rebase and miner updates

SergeyOvchinnik commented:
Made final changes, rebasing now

SergeyOvchinnik commented:
@motobob Please merge

motobob commented:
it now has conflicts
hm... let's close this, rebase over upstream and PR there directly.
